### PR TITLE
improve `ContainerRenderObjectMixin` error message when `parentData` is not set up properly

### PR DESCRIPTION
**Pull request by PurplePolyhedron**
Previously when subclassing `MultiChildRenderObjectWidget` and `RenderObject with ContainerRenderObjectMixin`, if one forgot to set up parent data,  the error message does not give hint that `setupParentData` need to be implemented by the `RenderObject`.

This PR add assertion that check parent data type before type cast and give hints if it is was not properly set.
## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/test/rendering/object_test.dart [hunk 1]
```diff
@@ -485,6 +491,16 @@ class TestRenderObject extends RenderObject with ContainerRenderObjectMixin<Test
   }
 }
 
+class TestRenderObjectWithoutSetupParentData extends TestRenderObject {
+  @override
+  void setupParentData(RenderObject child) {
+    // The default behavior for RenderObject.
```
Comment (chunhtai) [inline packages/flutter/test/rendering/object_test.dart:497]:
```suggestion
    // Use a mismatched parent data type
```

**Review by chunhtai — APPROVED** [review]
LGTM, just a nitpick

File: packages/flutter/test/rendering/object_test.dart [hunk 2]
```diff
@@ -377,6 +377,12 @@ void main() {
     root.buildScene(ui.SceneBuilder()).dispose();
     expect(calledBack, true);
   });
+
+  test('ContainerParentDataMixin asserts parentData type', () {
+    final TestRenderObject renderObject = TestRenderObjectWithoutSetupParentData();
+    final TestRenderObject child = TestRenderObject();
+    expect(()=>renderObject.add(child),throwsAssertionError);
```
Comment (navaronbracke) [inline packages/flutter/test/rendering/object_test.dart:384]: Consider matching the error message, using a `having()` selector on the `throwsAssertionError` matcher

Comment (navaronbracke) [inline packages/flutter/test/rendering/object_test.dart:384]:
```suggestion
    expect(() => renderObject.add(child), throwsAssertionError);
```

The formatting is wrong here.

**Review by dkwingsmt — APPROVED** [review]
LGTM, great change!

**Comment by PurplePolyhedron** [comment]
> LGTM, great change!

Thanks for the review. 
Can you add autosubmit? Or is there anything I need to do?

**Pull request merged**